File: crates/gitim-core/src/parser.rs
use crate::link::extract_links;
use crate::mention::extract_mentions;
use crate::types::{ChannelEvent, Handler, Message, ThreadEntry, ThreadFile};
use regex::Regex;
use std::sync::LazyLock;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ParseError {
    #[error("first non-empty line is not a message start (line {0})")]
    FirstLineNotMessage(usize),
    #[error("invalid handler in message at file line {line}: {source}")]
    InvalidHandler {
        line: usize,
        source: crate::types::handler::HandlerError,
    },
}

// Event tokens are kebab- or snake-case lowercase ASCII. Hyphens are
// part of the alphabet (e.g. `leave-workspace`) — the original `[a-z_]+`
// rejected anything compound and silently fell through the parser
// (offending lines were treated as continuation text).
static MSG_RE: LazyLock<Regex> = LazyLock::new(|| {
    crate::preconditions::regex_literal(
        r"^\[L(\d{6,})\]\[P(\d{6,})\]\[@([a-z0-9-]+)\]\[(\d{8}T\d{6}Z)\](?:\[E:([a-z][a-z0-9_-]*)\])? (.+)$",
    )
});

pub fn parse_thread(input: &str) -> Result<ThreadFile, ParseError> {
    let input = &input.replace("\r\n", "\n");
    if input.is_empty() {
        return Ok(ThreadFile { entries: vec![] });
    }

    let mut entries: Vec<ThreadEntry> = Vec::new();
    let mut current_body: Option<String> = None;
    let mut first_content_line = true;

    for (file_line_idx, line) in input.lines().enumerate() {
        if let Some(caps) = MSG_RE.captures(line) {
            finalize_entry(&mut entries, current_body.take());

            let line_number: u64 = crate::preconditions::parse_u64(&caps[1]);
            let point_to: u64 = crate::preconditions::parse_u64(&caps[2]);
            // `system` is rejected by `Handler::new` (it's reserved so no
            // user can claim it), but the daemon's cron engine emits
            // `[@system]` lines for cron fires. Reading those back has to
            // succeed — without this carve-out, parsing any thread file
            // containing a system message would fail. The factory
            // `Handler::system()` preserves the "no user-input path can
            // forge `system`" invariant; this parser just treats the
            // already-emitted token as the protocol-level constant.
            let raw_handler = &caps[3];
            let author = if raw_handler == "system" {
                Handler::system()
            } else {
                Handler::new(raw_handler).map_err(|e| ParseError::InvalidHandler {
                    line: file_line_idx + 1,
                    source: e,
                })?
            };
            let timestamp = caps[4].to_string();
            let event_type = caps.get(5).map(|m| m.as_str().to_string());
            let body_first_line = caps[6].to_string();

            if let Some(et) = event_type {
                entries.push(ThreadEntry::Event(ChannelEvent {
                    line_number,
                    point_to,
                    author,
                    timestamp,
                    event_type: et,
                    meta: serde_json::Value::Null,
                }));
            } else {
                entries.push(ThreadEntry::Message(Message {
                    line_number,
                    point_to,
                    author,
                    timestamp,
                    body: String::new(),
                    mentions: Vec::new(),
                    links: Vec::new(),
                }));
            }
            current_body = Some(body_first_line);
            first_content_line = false;
        } else {
            if first_content_line {
                return Err(ParseError::FirstLineNotMessage(file_line_idx + 1));
            }
            if let Some(ref mut body) = current_body {
                // Strip leading space escape if the remainder starts with [L (continuation rule)
                let content = if line.starts_with(" [L") {
                    &line[1..]
                } else {
                    line
                };
                body.push('\n');
                body.push_str(content);
            }
        }
    }

    finalize_entry(&mut entries, current_body.take());

    Ok(ThreadFile { entries })
}

fn finalize_entry(entries: &mut [ThreadEntry], body: Option<String>) {
    if let (Some(body), Some(entry)) = (body, entries.last_mut()) {
        match entry {
            ThreadEntry::Message(msg) => {
                msg.body = body;
                msg.mentions = extract_mentions(&msg.body);
                msg.links = extract_links(&msg.body);
            }
            ThreadEntry::Event(ev) => {
                ev.meta = serde_json::from_str(&body).unwrap_or(serde_json::Value::Null);
            }
        }
    }
}

File: products/cell/frontend/src/daemon-web/parser.ts
// TS port of gitim-core/src/parser.rs — parses .thread files into structured data.

export interface ParsedMessage {
  type: "message";
  line_number: number;
  point_to: number;
  author: string;
  timestamp: string;
  body: string;
}

export interface ParsedEvent {
  type: "event";
  line_number: number;
  point_to: number;
  author: string;
  timestamp: string;
  event_type: string;
  meta: Record<string, unknown>;
}

export type ThreadEntry = ParsedMessage | ParsedEvent;

export interface ThreadFile {
  entries: ThreadEntry[];
}

const MSG_RE =
  /^\[L(\d{6,})\]\[P(\d{6,})\]\[@([a-z0-9][a-z0-9-]*)\]\[(\d{8}T\d{6}Z)\](?:\[E:([^\]]+)\])?\s(.*)$/;

export function parseThread(text: string): ThreadFile {
  const input = text.replace(/\r\n/g, "\n");
  if (input.length === 0) {
    return { entries: [] };
  }

  const entries: ThreadEntry[] = [];
  let currentBody: string | null = null;

  // Rust's str::lines() doesn't yield a trailing empty string for a final '\n'.
  // JS split("\n") does, which would add a spurious continuation line. Match Rust behavior.
  const lines = input.endsWith("\n") ? input.slice(0, -1).split("\n") : input.split("\n");

  for (const line of lines) {
    const match = MSG_RE.exec(line);

    if (match) {
      // Finalize previous entry before starting a new one
      finalizeEntry(entries, currentBody);

      const lineNumber = parseInt(match[1], 10);
      const pointTo = parseInt(match[2], 10);
      const author = match[3];
      const timestamp = match[4];
      const eventType = match[5] ?? null;
      const bodyFirstLine = match[6];

      if (eventType !== null) {
        entries.push({
          type: "event",
          line_number: lineNumber,
          point_to: pointTo,
          author,
          timestamp,
          event_type: eventType,
          meta: {},
        });
      } else {
        entries.push({
          type: "message",
          line_number: lineNumber,
          point_to: pointTo,
          author,
          timestamp,
          body: "",
        });
      }

      currentBody = bodyFirstLine;
    } else {
      // Continuation line — append to current body
      if (currentBody !== null) {
        // Strip leading space escape for lines that look like message prefixes (spec 5.3 rule 5)
        const content = line.startsWith(" [L") ? line.slice(1) : line;
        currentBody += "\n" + content;
      }
    }
  }

  // Finalize the last entry
  finalizeEntry(entries, currentBody);

  return { entries };
}

function finalizeEntry(
  entries: ThreadEntry[],
  body: string | null,
): void {
  if (body === null || entries.length === 0) return;

  const entry = entries[entries.length - 1];
  if (entry.type === "message") {
    entry.body = body;
  } else {
    try {
      entry.meta = JSON.parse(body) as Record<string, unknown>;
    } catch {
      entry.meta = {};
    }
  }
}
